### ISSUE TITLE
Generalize Request/Response

### ISSUE THREAD

**Issue by candlejack297** [scrapy/scrapy#2182]
It would be nice if scrapy had a slightly more general abstraction over request/responses. 

My use case involves a Rest API that has a python sdk. Instead of using scrapy to make the direct http requests/responses, I'd like to make the request using the sdk, but then follow the normal scrapy process (middleware, parse, item pipeline, etc). I can and have made this happen with a few custom classes, but it still feels like there should be an easier way to hook things into the pipeline (maybe there is, and I'm just missing it). I realize it's not exactly crawling to do something like this, but we have some "normal" spiders as well, and it's nice to have a single entry point for all our scraping.

**Comment by kmike** [comment]
Hey @candlejack297,

What the API could look like? Could you provide an example?

Can you make these requests e.g. in start_requests, parse the response and send scrapy Requests from there?

A problem with third-party libraries which make HTTP requests is that usually they do that using urllib or requests; it means these requests block the event loop. You can use these libraries with Scrapy, but requests won't be sent in parallel.

**Comment by candlejack297** [comment]
I guess a naive API would be something like Action/Response (without the http trappings) where Request/Response could just subclass them. I was envisioning a scenario where scrapy essentially invokes a custom function in place of Requests and lets you process them in place of Responses. Blocking the event loop would be problematic, but I was thinking scrapy would parallelize the custom function invocation?

**Comment by wRAR** [comment]
> Instead of using scrapy to make the direct http requests/responses, I'd like to make the request using the sdk

This can be done in custom download handlers. Otherwise I don't see what can be done (or discussed) here as the proposal is too vague.

**Comment by Gallaecio** [comment]
https://github.com/scrapy-plugins/scrapy-zyte-api is a good example of this, if anyone wants to do something similar.

**Comment by wRAR** [comment]
https://github.com/scrapy-plugins/scrapy-playwright too.
